torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1662,7 +1682,31 @@ struct to_ir {
         graph->createStore(x, fv)->insertBefore(false_block->return_node());
       }
 
-      auto unified = unifyTypes(tv->type(), fv->type());
+      // If we already know that the type is a Union, then we want to
+      // allow heterogeneous types to be unified
+      c10::optional<TypePtr> full_type = environment_stack->getType(x);
```

Comment:
So, the intent of this logic is to allow:
```
def foo(x: Union[int, str]):
    if cond:
        x = 2
    else:
        x = "str"
   return x
```
but not allow
 ```
def foo():
    if cond:
        x = 2
    else:
        x = "str"
   return x
```

Could you explain the intention behind this?

Reply:
Yes, that's exactly what I'm trying to do. If we allowed the latter example, then I think we'd fail to catch some legitimate errors. I like the idea of only allowing `Union` if the user explicitly annotates it

cc @gmagogsfm